phoenix-core-client/src/main/java/org/apache/phoenix/jdbc/HighAvailabilityPolicy.java

```diff
@@ -123,15 +158,88 @@ public Connection provide(HighAvailabilityGroup haGroup, Properties info,
+                               ClusterRoleRecord newRecord) throws SQLException {
+            if (!Objects.equals(oldRecord.getUrl1(), newRecord.getUrl1()) &&
+                    !Objects.equals(oldRecord.getUrl1(), newRecord.getUrl2())) {
+                LOG.info("Cluster {} is changed to {} in HA group {}, now closing all its connections",
+                        oldRecord.getUrl1(), newRecord.getUrl1(), haGroup);
+                closeConnections(haGroup, oldRecord.getUrl1(), oldRecord.getRegistryType());
+            }
+            if (!Objects.equals(oldRecord.getUrl2(), newRecord.getUrl2()) &&
+                    !Objects.equals(oldRecord.getUrl2(), newRecord.getUrl1())) {
+                LOG.info("Cluster {} is changed to {} in HA group {}, now closing all its connections",
+                        oldRecord.getUrl2(), newRecord.getUrl2(), haGroup);
+                closeConnections(haGroup, oldRecord.getUrl2(), oldRecord.getRegistryType());
+            }
+
         }
     };
 
     private static final Logger LOG = LoggerFactory.getLogger(HighAvailabilityGroup.class);
 
+    /**
+     * Utility to close cqs and all it's connection for specific url of a HAGroup
+     * @param haGroup The High Availability (HA) Group
+     * @param url The url on which cqs and connections are present
+     * @param registryType The registry Type of connections
+     * @throws SQLException if fails to close the connections
+     */
+    private static void closeConnections(HighAvailabilityGroup haGroup, String url,
+                                         ClusterRoleRecord.RegistryType registryType) throws SQLException {
+        ConnectionQueryServices cqs = null;
+        //Close connections for every HAURLInfo's (different principal) conn for a give HAGroup
+        for (HAURLInfo haurlInfo : HighAvailabilityGroup.URLS.get(haGroup.getGroupInfo())) {
+            try {
+                cqs = PhoenixDriver.INSTANCE.getConnectionQueryServices(
+                        HighAvailabilityGroup.getJDBCUrl(url, haurlInfo, registryType), haGroup.getProperties());
+                cqs.closeAllConnections(new SQLExceptionInfo
+                        .Builder(SQLExceptionCode.HA_CLOSED_AFTER_FAILOVER)
+                        .setMessage("Phoenix connection got closed due to failover")
+                        .setHaGroupInfo(haGroup.getGroupInfo().toString()));
+                LOG.info("Closed all connections to cluster {} for HA group {}", url,
+                        haGroup.getGroupInfo());
+            } finally {
+                if (cqs != null) {
+                    // CQS is closed, but it is not invalidated from global cache in PhoenixDriver
+                    // so that any new connection will get error instead of creating a new CQS
+                    LOG.info("Closing CQS after clusterRoleRecord change for '{}'", url);
+                    cqs.close();
```

Comment:
Will also need to remove it from the CQSI map, otherwise the map will contain a closed CQSI and will never initialize again.